## Consideraciones iniciales

A través de la herramienta de Probabilidad de baja por formulario buscamos ofrecer una propuesta al cliente que incremente las posibilidades de permanecer en la compañia.<br>
La herramienta se agrega a los procedimientos de atención habituales y realiza recomendaciones en función al valor de probabilidad. Los rangos quedarán definidos por los costos, la capacidad disponible y la estrategia que se quiera adoptar.<br>
La argumentación será cuantitativa en partes y cualitativa en otras dado que es un ejercicio hipotético y no poseemos datos adicionales/lineamientos que terminarían de definir los escenarios.<br>
Recordemos, como vimos en el análisis exploratorio, que los motivos se engloban dentro de los tópicos: "Competidor" con el 33% del total (El competidor ofrece mejor servicio u oferta) y por otro lado la "Atención" (El Soporte y Actitud brindados) acumula el 20% de los casos. Las acciones irán apuntadas en esas direcciones y no debemos confundir los atributos del modelo que potencian la probabilidad de baja, con las causas.

## Ideas principales sobre la solución

Los motivos de baja enumerados anteriormente permiten que los equipos responsables de la oferta comercial, tecnología, capacitación y recursos humanos actúen con soluciones estruturales.<br>
Nuestro enfoque estará orientado a utilizar los recursos actuales para ofrecer un mejor escenario en el corto plazo reduciendo la propensión a la baja.

**Servicio y oferta**<br>
Recordemos respecto al servicio que la mayor tasa de churn se lo lleva "Fiber optic" con un 42%, luego "DSL" 19% y 7% los que no tienen internet. Fibra óptica es  la mejor tecnología disponible en la empresa para los clientes y la mejora dependerá de soluciones de infraestructura de mediano y largo plazo. No enfocaremos acciones en este departamento.<br>
Respecto a oferta podemos pensar, propuestas de valor, para el corto y mediano plazo, que permitan retener clientes a la espera de una mejor propuesta del áreas involucradas. Un porcentaje de descuento en el abono actual y/o servicios bonificados serán nuestras 2 alternativas consideradas.

**Atención**<br>
Aquí la estrategia es emparchar la atención de insuficiente calidad en primera línea de atención técnica con la atención de "Tech Support", recordemos que los que no cuentan con este servicio mostraron un churn del 42% contra un 15% de los que cuentan con él. Suponemos que esta diferencia se fundamenta en 3 ejes:la calidad de atención, la formación y las herramientas.<br><br>

Nota: Nos enfocaremos en la gestión sobre consultas/reclamos que ingresen por el canal técnico, pero podríamos aplicar una solución simular para el canal comercial.

## Importación y preparación de los datos

In [ ]:
#Importación librerías
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
#Importación dataset desde github 
url='https://raw.githubusercontent.com/marcelobour/telco_churn/main/data/Telco_customer_churn.csv'
telco = pd.read_csv(url, sep=';')

#Definimos una lista con las columnas a convertir a número
cols = ['Longitude', 'Latitude', 'Monthly Charges']

#Reemplazamos la coma por punto y el espacio por nada y convertimos a número
telco[cols] = telco[cols].apply(lambda x: x.str.replace(',', '.')).apply(lambda x: x.str.replace(' ', '0')).apply(lambda x: pd.to_numeric(x))

#Visualizamos los datos
pd.options.display.max_columns = None
display(telco.head())


CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   
2  9305-CDSKC      1  United States  California  Los Angeles     90006   
3  7892-POOKP      1  United States  California  Los Angeles     90010   
4  0280-XJGEX      1  United States  California  Los Angeles     90015   

                 Lat Long   Latitude   Longitude  Gender Senior Citizen  \
0  33.964131, -118.272783  33.964131 -118.272783    Male             No   
1   34.059281, -118.30742  34.059281 -118.307420  Female             No   
2  34.048013, -118.293953  34.048013 -118.293953  Female             No   
3  34.062125, -118.315709  34.062125 -118.315709  Female             No   
4  34.039224, -118.266293  34.039224 -118.266293    Male             No   

  Partner Dependents  Tenure Months Phone Service Multiple Lines  \
0      No         No              2           Yes             No   
1      No        Yes              2           Yes             No   
2      No        Yes              8           Yes            Yes   
3     Yes        Yes             28           Yes            Yes   
4      No        Yes             49           Yes            Yes   

  Internet Service Online Security Online Backup Device Protection  \
0              DSL             Yes           Yes                No   
1      Fiber optic              No            No                No   
2      Fiber optic              No            No               Yes   
3      Fiber optic              No            No               Yes   
4      Fiber optic              No           Yes               Yes   

  Tech Support Streaming TV Streaming Movies        Contract  \
0           No           No               No  Month-to-month   
1           No           No               No  Month-to-month   
2           No          Yes              Yes  Month-to-month   
3          Yes          Yes              Yes  Month-to-month   
4           No          Yes              Yes  Month-to-month   

  Paperless Billing             Payment Method  Monthly Charges Total Charges  \
0               Yes               Mailed check            53.85        108,15   
1               Yes           Electronic check            70.70        151,65   
2               Yes           Electronic check            99.65         820,5   
3               Yes           Electronic check           104.80       3046,05   
4               Yes  Bank transfer (automatic)           103.70        5036,3   

  Churn Label  Churn Value  Churn Score  CLTV                   Churn Reason  
0         Yes            1           86  3239   Competitor made better offer  
1         Yes            1           67  2701                          Moved  
2         Yes            1           86  5372                          Moved  
3         Yes            1           84  5003                          Moved  
4         Yes            1           89  5340  Competitor had better devices

## Impacto en el churn al cambiar contrato mensual a anual y agregar Tech Support

El modelo reducido de 8 variables, que demostramos tiene valores de precicisión y recall apenas por debajo del modelo completo, presenta los siguientes como determinantes de la propensión a la baja:
*   Tenure Months
*   Monthly Charges
*   Latitude
*   Longitude
*   Dependents
*   Contract
*   Paperless Billing
*   Internet Service

Respecto a 'Tenure Months' en líneas generales cuantos más meses en la compañía menor propensión a la baja, los clientes que permanecen entre 1 y 5 meses tienen 54% de churn (el doble de la media de la compañia).<br>
En cuanto a Contract los clientes con mensual tienen 43% de churn, mientras que los anuales 11%, y los bianuales 3%, si bien esto no es una causa puede ayudarnos a reducir las bajas ofreciendo promociones por cambio de tipo de contrato.
El modelo entrenado reconoce o valoriza más alta la propensión a la baja cuando las personas no tienen gente a cargo ('Dependents'), cuando cuentan con fibra óptica ('Internet Service') y cuando la factura es digital ('Paperless Billing').

Veamos como se relacionan algunos escenarios de estas variables con Tech Support:

In [ ]:
df = telco.loc[(telco['Dependents']=='No')& 
          (telco['Internet Service']=='Fiber optic')&
          (telco['Paperless Billing']=='Yes')]
data = pd.pivot_table(df, values='Churn Value', index=['Contract'], columns=['Tech Support'], aggfunc=np.mean).round(2) 
display(data)

Tech Support      No   Yes
Contract                  
Month-to-month  0.64  0.44
One year        0.24  0.27
Two year        0.14  0.11

Nuestras acciones estarán enfocadas, según corresponda, a ofrecer Tech Support, convertir los contratos a anuales, y ofrecer bonificaciones/descuentos.<br>
El detalle de los procedimientos https://docs.google.com/presentation/d/1eqBfX3qzZ_CABcjgoiHDySgU-elu64DJCei-NLWb4uM/edit#slide=id.p

## Captura de la retención preventiva

Vamos a suponer algunos valores que podríamos haber obtenido consultando a referentes de la compañia o que podríamos haber estimado a partir de una prueba piloto.<br>
*  Probabilidad de contacto con motivos técnicos mensual: 70%.
*  Índice de reitero: 60%.
*  Fracción de los casos técnicos que puede atender la isla de Tech Support: 40%.
*  Capacidad ociosa media de la isla de Tech Support: 30% promedio. 
*  Capacidad ociosa media deseada de la isla de Tech Support: 10% promedio.
*  Clientes que mencionan que van a dar de baja el servicio: 1 de cada 3.
*  Clientes que aceptan el ofrecimiento de Tech Support bonificado por 1 mes + contrato anual = 30%.

De los datos anteriores se deprende que existe una capacidad del 5% de la isla de Tech Support utilizable para nuestra implementación sin generar costos adicionales.

In [ ]:
#Cuántos clientes activos tenemos?
activos = telco.loc[telco['Churn Value']==0].shape[0]
activos

5174

In [ ]:
#Cúantos contactos por motivos técnicos?
motivo_tec = 0.7
contac_tec = round(activos * motivo_tec)
contac_tec

3622

In [ ]:
#Cuántos tienen Tech Support contratado?
with_tech_support = telco.loc[telco['Tech Support']=='Yes'].shape[0]
percen_with_tech_support = round(with_tech_support/activos, 2)
percen_with_tech_support

0.4

In [ ]:
#Cuántos se derivan a Tech Support?
competencia_tech_support = 0.4
derivados_tech_support = round(contac_tec * competencia_tech_support * percen_with_tech_support)
derivados_tech_support

580

In [ ]:
#Cuánta capacidad, en contactos mensuales, podemos aprovechar?
capactual = 0.7
capmax = 0.9
contactos_sin_costo = round(derivados_tech_support / capactual * capmax - derivados_tech_support)
contactos_sin_costo

166

In [ ]:
#Cuántos clientes únicos representa?
reitero = 0.6
casos_sin_costo = round(contactos_sin_costo / (1 + reitero))
casos_sin_costo

104

Por lo tanto nuestro objetivo es capturar 104 clientes mensualmente con la mayor probabilidad de baja, que no tengan tech support contratado ni contrato anual. A estos realizar el ofrecimiento de la bonificación del servicio por un mes y la derivación inmediata si accede a un contrato anual con un 10% de descuento.

In [ ]:
#Cuántos casos ingresan mensualmente sin Tech Support y contrato mensual
contactos_pool = round(telco.loc[(telco['Tech Support']=='No') & (telco['Contract']=='Month-to-month')].shape[0] * motivo_tec)
contactos_pool

1876

In [ ]:
#Cuántos clientes únicos representan esos contactos
clientes_pool = round(contactos_pool / (1 + reitero))
clientes_pool

1172

In [ ]:
#Qué porcentaje debemos capturar de los que ingresa?
recall_objetivo = round(casos_sin_costo / clientes_pool, 2)
recall_objetivo

0.09

In [ ]:
#A qué porcentaje le deberíamos ofrecer sabiendo que sólo el 30% acepta?
efectividad_ofrecimiento = 0.3
recall_objetivo_ofrecimiento = round(recall_objetivo / efectividad_ofrecimiento, 2)
recall_objetivo_ofrecimiento

0.3

A partir de los valores que habíamos analizado en: <br> https://colab.research.google.com/drive/1mAljIO4qR_3l-ufKnFnwRDu70YZ7J7I0 <br> utilizamos la tabla que relaciona recall, precision y threshold:

In [ ]:
#Importación datos del modelo elegido de 8 variables desde github 
url_data_8var='https://raw.githubusercontent.com/marcelobour/telco_churn/main/data/aucpr-precision-8var.csv'
data_8var = pd.read_csv(url_data_8var)
display(data_8var)

Unnamed: 0  re_avg  pre_avg    re_std   pre_std  re_rel_error  \
0           0    0.58     0.69  0.017840  0.012149      0.060288   
1           1    0.49     0.72  0.019678  0.015916      0.078714   
2           2    0.41     0.75  0.018495  0.016709      0.088416   
3           3    0.30     0.81  0.017499  0.025042      0.114325   
4           4    0.23     0.88  0.016761  0.029449      0.142830   
5           5    0.14     0.90  0.015421  0.036141      0.215898   
6           6    0.08     0.95  0.011885  0.034032      0.291194   
7           7    0.03     0.96  0.006477  0.044851      0.423177   
8           8     NaN     0.67  0.001826  0.479463           inf   
9           9     NaN      NaN       NaN       NaN           NaN   

   pre_rel_error  thresh  
0       0.034509    0.50  
1       0.043328    0.55  
2       0.043667    0.60  
3       0.060597    0.65  
4       0.065591    0.70  
5       0.078708    0.75  
6       0.070213    0.80  
7       0.091571    0.85  
8       1.402609    0.90  
9            NaN    0.95

Con esta tabla podemos relacionar el thresh (probabilidad de baja) con el recall (captura).<br>

Si sólo nos guiáramos por la tabla anterior y no tuvieramos en cuenta que la intención manifiesta de baja aumenta las probabilidades entonces elegiríamos un threshold de 0.65 que nos brinda un recall de 30% (fila 3 de la tabla).<br> 
Pero como indicamos anteriormente, sabemos que 1 de cada 3 clientes manifiesta la intención de baja y que cuando lo hace y la predicción de baja es positiva (thresh >= 0.5) la probabilidad asciende y también deseamos capturarlos por lo que debemos capturar en 2 cuotas:<br>
*  A los que no manifiestan baja (2/3) los capturamos por su valor de probabilidad. Es decir el 20%, una probabilidad de baja del 70% nos brinda aproximadamente ese recall, 23% (fila 4 de la tabla).  A este grupo nos referimemos como "alta probabilidad".<br>
*  A los que manifiestan la baja los seleccionamos por un valor de probabilidad que nos entregue justamente el tercio restante, otro 10% de recall. Como ya tenemos el 23% por la fila 4, necesitamos 7%, y como en realidad 1 de cada 3 van a manifestar la baja, deberíamos conseguir un recall adicional del triple es decir 21%. 23% + 21% = 44%, por aproximación y para ser conservadores con la capacidad de la isla Tech Support (considerando que tenemos un error del 14% y 9% respectivamente) seleccionamos la probabilidad 0.6 que no totaliza un 41% de recall (fila 2 de la tabla). A este grupo nos referiremos como "manifestantes".


Esto resulta en un nivel de probabilidad crítico entre 0.7 y 1 dentro del cual realizaremos el ofrecimiento siempre (Tech support bonificado por 1 mes + oferta contrato anual) y otro nivel, de 0.6 a 0.7, donde lo haremos sólo si manifiesta intención de baja (Sólo Tech support por única vez). Por debajo de 0.6 no ofreceremos cambios en el procedimiento desde esta solución de retención preventiva.

## Alcance e Impacto

In [ ]:
#Grupo "alta probabilidad"
cuota_alta_prob = 0.23
precision_alta_prob = 0.88

#Grupo "manifestantes"
cuota_manif = (0.41 - 0.23)/3
precision_manif = 0.75

#Clientes alcanzados
alcanzados_total = round(clientes_pool * efectividad_ofrecimiento * (cuota_alta_prob + cuota_manif), 0)
alcanzados_total

102.0

Recordemos que teníamos un objetivo de 104, prácticamente alcanzado.

Para calcular la precisión final, es decir la precisión ponderada resultante debemos componer las precisiones de ambas cuotas:<br>

In [ ]:
#Promedio ponderado de las precisiones
precision_final = round(((cuota_alta_prob * precision_alta_prob) + (cuota_manif * precision_manif)) / (cuota_alta_prob + cuota_manif), 2)
precision_final

0.85

Es decir que el 85% de los casos que aceptan el ofrecimiento efectivamente se habrían dado de baja.<br>
Pero cual es el costo de ofrecer una bonificación/oferta a alguien que finalmente no hubiera dado de baja?
El grupo manifestante no genera ningún costo, ofrecemos transferencia a Tech Support por única vez y eso lo capturamos de la capacidad ociosa disponible.
El grupo de alta probabilidad que no habría dado de baja (15%) genera un costo (reducción de 10% en la tarifa).

In [ ]:
#Participación de las cuota alta probabilidad
part_cuota_alta_prob = round(cuota_alta_prob/(cuota_alta_prob + cuota_manif), 2)
part_cuota_alta_prob

0.79

Es decir el 15% del 79% nos genera una disminución en el abono (descuento 10% anual. Pero eso lo cubrimos con el 85% (del 100%) de los casos que aceptan el ofrecimiento.

A partir de aquí ya sería necesario tener resultados preliminares o de un piloto para calcular repago y evaluar si ofrecer un mayor o menor descuento podrían ofrecernos mejores escenarios. Tenemos 2 grandes incertidumbres:<br>
*  Cuál es la tasa de baja de los clientes que retuvimos preventivamente?
*  La interacción de esta acción con otras acciones estructurales que ofrezca la empresa como mejorar la atención en primera línea u ofrecer paquetes más competitivos proactivamente.

En torno a la primer pregunta podríamos partir de la hipótesis (basados en la tabla de bloque [4]) que con la acción bajamos el churn:
*  Un 60% aprox para el grupo "alta probabilidad" (De 64% de churn 24%)
*  Un 30% aprox para el grupo "manifestantes" (De 64% a 44%)


Por lo cual tendríamos la siguiente balanza final: 

In [ ]:
#No va a dar de baja pero le aplicamos descuento 10%
descuento = 0.10
costo = part_cuota_alta_prob * (1-precision_final) * descuento

#Acepta descuento pero redujimos un 60% las chances de baja (Grupo: Alta probabilidad)
ganancia_alta_prob = part_cuota_alta_prob * precision_final * (1-descuento) * 0.60 

#Acepta transferencia a Tech (Grupo: Manifestantes)
ganancia_manif = (1-part_cuota_alta_prob) * 0.30

neto = round(ganancia_alta_prob + ganancia_manif - costo, 2)
neto

0.41

Es decir que con esta acción logramos retener el 41% de los abonos de las personas que acepten el ofrecimiento que corresponda según escenario. <br>
Una variente para la implementación consistiría en ampliar la capacidad de Tech Support analizando los costos de demanda mensual por cliente y margen de ganancia en el abono y multiplicar el alcance.